### PR TITLE
LibWasm: Sync!

### PR DESCRIPTION
**Pull request by alimpfard**
It's wasm sync time!

### CODE REVIEW

**Comment by nico** [comment]
`==23918==ERROR: AddressSanitizer: heap-use-after-free on address 0x612000481cc8 at pc 0x000100e6b6f8 bp 0x00016fb5ca90 sp 0x00016fb5ca88` on Mac CI looks real and related.

**Comment by alimpfard** [comment]
🤔 I think this is a repeat of <https://github.com/LadybirdBrowser/ladybird/actions/runs/9977375108/job/27571834471?pr=626#step:15:46>, hopefully it stays put so I can actually fix it this time.

**Comment by alimpfard** [comment]
Tried a bunch of times, can't repro locally this time either.
Let's keep the PR open until I figure it out though.

**Comment by ADKaster** [comment]
It showed up here as well <https://github.com/LadybirdBrowser/ladybird/actions/runs/10087559296/job/27891908376?pr=821#step:17:46>

File: AK/SIMDExtras.h [hunk 1]
```diff
@@ -213,13 +213,14 @@ ALWAYS_INLINE static T shuffle_or_0_impl(T a, Control control, IndexSequence<Idx
...
         // So we jinx its __builtin_shuffle to work with out of bounds indices
+        // TODO: verify that this masking logic is correct (for machines with __builtin_shuffle)
         auto mask = (control >= 0) | (control < N);
         return __builtin_shuffle(a, control & mask) & ~mask;
     }
     // 1. Set all out of bounds values to ~0
     // Note: This is done so that  the optimization mentioned down below works
     // Note: Vector compares result in bitmasks, aka all 1s or all 0s per element
-    control |= ~((control > 0) | (control < N));
+    control |= ~((control >= 0) & (control < N));
```
Comment (Hendiadyoin1) [inline AK/SIMDExtras.h:223]: Oops, how did that happen

File: AK/SIMDExtras.h [hunk 2]
```diff
@@ -213,13 +213,14 @@ ALWAYS_INLINE static T shuffle_or_0_impl(T a, Control control, IndexSequence<Idx
     if constexpr (__has_builtin(__builtin_shuffle)) {
         // GCC does a very bad job at optimizing the masking, while not recognizing the shuffle idiom
         // So we jinx its __builtin_shuffle to work with out of bounds indices
+        // TODO: verify that this masking logic is correct (for machines with __builtin_shuffle)
```
Comment (Hendiadyoin1) [inline AK/SIMDExtras.h:216]: This should be any GCC build, which is the default for many people

File: Userland/Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp [hunk 3]
```diff
@@ -1289,12 +1289,17 @@ void BytecodeInterpreter::interpret(Configuration& configuration, InstructionPoi
...
+        auto& arg = instruction.arguments().get<Instruction::ShuffleArgument>();
+        auto b = *pop_vector<u8, MakeUnsigned>(configuration);
+        auto a = *pop_vector<u8, MakeUnsigned>(configuration);
+        using VectorType = Native128ByteVectorOf<u8, MakeUnsigned>;
+        VectorType result;
+        for (size_t i = 0; i < 16; ++i)
+            if (arg.lanes[i] < 16)
+                result[i] = a[arg.lanes[i]];
+            else
+                result[i] = b[arg.lanes[i] - 16];
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp:1278]: quick question, is the index enforced to be in the range of 0-32?

**Review by Hendiadyoin1 — COMMENTED** [review]
Was on vacation, so I did not see the OG PRs
but here are a few things I have noticed

File: Userland/Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp:1278]: Yes.
https://github.com/SerenityOS/serenity/blob/568bd04b610bdae14fb37ccfae11788fe32261c3/Userland/Libraries/LibWasm/AbstractMachine/Validator.cpp#L2352

**Comment by nico** [comment]
Here too, so seems to predate this PR: https://github.com/SerenityOS/serenity/actions/runs/10135938296/job/28024018883?pr=24876

(Ie merging this is probably fine)

**Comment by alimpfard** [comment]
Alright, will update PR and merge later today.

**Comment by circl-lastname** [comment]
For the future, is it possible to link the synced PRs from Ladybird like so LadybirdBrowser/ladybird#<pr num>? Makes it easier to tell from the PR page.

**Comment by alimpfard** [comment]
oops, forgot to `cherry-pick -x` :P

**Comment by alimpfard** [comment]
> For the future, is it possible to link the synced PRs from Ladybird like so `LadybirdBrowser/ladybird#<pr num>`? Makes it easier to tell from the PR page.

That seems like a bunch of manual work?

**Comment by circl-lastname** [comment]
Syncs at least:
LadybirdBrowser/ladybird#794
LadybirdBrowser/ladybird#811
LadybirdBrowser/ladybird#850
LadybirdBrowser/ladybird#874
LadybirdBrowser/ladybird#875

**Comment by circl-lastname** [comment]
Thanks :^)
About the manual work part, I've put up a `gh` (the GitHub CLI) based [script](https://github.com/circl-lastname/LBSync) that can do most of the syncing work automatically on my profile.

**Comment by nico** [comment]
Also:
https://github.com/LadybirdBrowser/ladybird/pull/671
https://github.com/LadybirdBrowser/ladybird/pull/702
https://github.com/LadybirdBrowser/ladybird/pull/719

**Comment by nico** [comment]
…and one commit from https://github.com/LadybirdBrowser/ladybird/pull/851

That's all, together with the ones mentioned in https://github.com/SerenityOS/serenity/pull/24814#issuecomment-2255737061

**Pull request merged**